BRUCE-6721 - Implement a queue for tasks to perform redis DB operations

AngithCJ opened the pull request:
**Changes**

- Implemented a queue for tasks to perform update task operation

AngithCJ committed:
Implement a job queue for tasks to perform update task operation

AngithCJ committed:
Added unit tests, code review changes

AngithCJ committed:
file format is changed to LF from CRLF

AngithCJ committed:
1. Implemented the task with new approach as to avoid double copy from the implementation
2. Fixed and added unit tests

AngithCJ committed:
Fixed the synchronization issues between goroutines

Bharath-KKB commented on lib-persistence-manager/persistencemgr/redis.go:
```diff
@@ -663,6 +670,99 @@ func (p *ConnPool) SaveBMCInventory(data map[string]interface{}) *errors.Error {
 }
 
+// GetWriteConnection retrieve a write connection from the connection pool
+func (p *ConnPool) GetWriteConnection() (*Conn, *errors.Error) {
```
we already have this method dont we?

AngithCJ replied:
No. We were retrieving the connection in each function.

AngithCJ replied:
Made the changes as discussed in the call.

Bharath-KKB commented on lib-persistence-manager/persistencemgr/redis.go:
```diff
@@ -663,6 +670,99 @@ func (p *ConnPool) SaveBMCInventory(data map[string]interface{}) *errors.Error {
+
+// InitRedisPipeline starts a pipelined transaction
+func (c *Conn) InitRedisPipeline() error {
+	if c.WriteConn == nil || c.WritePool == nil {
```
How is this handled upstream ?

AngithCJ replied:
This method is removed as we are sending tasks in a map now. Errors are handled as we discussed in the call.

Bharath-KKB commented on lib-persistence-manager/persistencemgr/redis.go:
```diff
@@ -663,6 +670,99 @@ func (p *ConnPool) SaveBMCInventory(data map[string]interface{}) *errors.Error {
+	}
+	jsondata, marshalErr := json.Marshal(val)
+	if marshalErr != nil {
+		c.WriteConn.Send("DISCARD")
```
why discard if one json fails ? And why close the connection ?

AngithCJ replied:
can we avoid the error and continue with the other requests?

AngithCJ replied:
made the changes. if JSON marshal is failed it would be logged and continue the operation.

Bharath-KKB commented on lib-persistence-manager/persistencemgr/redis.go:
```diff
@@ -663,6 +670,99 @@ func (p *ConnPool) SaveBMCInventory(data map[string]interface{}) *errors.Error {
+	}
+	_, createErr := c.WriteConn.Do("SET", key, jsondata)
+	if createErr != nil {
+		c.WriteConn.Send("DISCARD")
```
Same here

AngithCJ replied:
This is probably a DB issue, right? Shouldn't we stop the execution here also?

AngithCJ replied:
Implemented retry to handle these errors. returning DB update failed error and once this error happens, I will be retrying the operation again.

Bharath-KKB commented on lib-persistence-manager/persistencemgr/redis.go:
```diff
@@ -663,6 +670,99 @@ func (p *ConnPool) SaveBMCInventory(data map[string]interface{}) *errors.Error {
+	}
+	_, createErr := c.WriteConn.Do("ZADD", index, value, key)
+	if createErr != nil {
+		c.WriteConn.Send("DISCARD")
```
Same here

AngithCJ replied:
did the same error handling here as well

Bharath-KKB commented on lib-persistence-manager/persistencemgr/redis.go:
```diff
@@ -663,6 +670,99 @@ func (p *ConnPool) SaveBMCInventory(data map[string]interface{}) *errors.Error {
+	if c.WriteConn == nil || c.WritePool == nil {
+		return errors.PackError(errors.UndefinedErrorType, "InitRedisPipeline :  WriteConn OR WritePool is nil")
+	}
+	defer c.Close()
```
This is implicitly closign connection. The caller should close the connectio n .

AngithCJ replied:
Ok. will make the changes

AngithCJ replied:
made the changes

Bharath-KKB commented on lib-persistence-manager/persistencemgr/redis_test.go:
```diff
@@ -30,6 +30,39 @@ import (
 	"github.com/gomodule/redigo/redis"
 )
 
+type MockConn struct {
```
where do we actually test the DB connection codes ?

AngithCJ replied:
From here, I have added the tests https://github.com/AngithCJ/ODIM/pull/34/files#diff-771ec6af6c5d0e1d35ed5294a390a2ce8aa425a490080e9e250027c097fe296fR1453

Bharath-KKB commented on svc-task/main.go:
```diff
@@ -77,6 +79,10 @@ func main() {
 	}
 
+	queueSize := 1000
+	dbCommitInterval := time.Millisecond
```
This should not be hard coded

AngithCJ replied:
Ok.

AngithCJ replied:
This is configured using odimra config

Bharath-KKB commented on svc-task/thandle/thandle.go:
```diff
@@ -53,7 +53,7 @@ type TasksRPC struct {
 	DeleteTaskFromDBModel            func(t *tmodel.Task) error
 	DeleteTaskIndex                  func(taskID string) error
-	UpdateTaskStatusModel            func(t *tmodel.Task, db common.DbType) error
+	UpdateTaskStatusModel            func(t *tmodel.Task)
```
This updates the queue isnt it? So why is it not part of name ?

AngithCJ replied:
This was the previous method. I did not change the name. I will make the changes.

AngithCJ replied:
changed the name to UpdateTaskQueue

Bharath-KKB commented on svc-task/thandle/thandle.go:
```diff
@@ -365,22 +365,14 @@ func (ts *TasksRPC) taskCancelCallBack(taskID string) error {
 			ts.DeleteTaskFromDBModel(subTask)
 		} else if subTask.TaskState != common.Cancelling {
 			subTask.TaskState = common.Cancelling
-			err := ts.UpdateTaskStatusModel(subTask, common.InMemory)
```
who has reviewed previous commits?

Bharath-KKB commented on svc-task/thandle/thandle.go:
```diff
@@ -885,19 +877,11 @@ func (ts *TasksRPC) CreateChildTaskUtil(userName string, parentTaskID string) (s
-		l.Log.Error("error while updating the child/sub task details in to DB: " + err.Error())
-		return "", fmt.Errorf("error while updating the child/sub task details: " + err.Error())
-	}
+	ts.UpdateTaskStatusModel(childTask)
```
why is this not handling error ?

AngithCJ replied:
UpdateTaskStatusModel adds the task to the channel (queue). It does not return any error.

Bharath-KKB commented on svc-task/tmessagebus/publish.go:
```diff
@@ -61,7 +61,7 @@ func Publish(taskURI, messageID, eventType, taskMessage string) {
 
 	if err := k.Distribute(mbevent); err != nil {
 		l.Log.Error("unable to publish the event to message bus: " + err.Error())
+		l.Log.Error("info: TaskURI:" + taskURI + ", EventID:" + eventID + ", MessageID:" + messageID)
```
Why is info: part of the message ??? Also should be one call to Log.Error

AngithCJ replied:
It is a previous log that was outside of the if block. It was causing a lot of logs in the task service. I only moved that from outside to the if block to reduce the logs. I will remove the info from the log.

AngithCJ replied:
made the changes

Bharath-KKB commented on svc-task/tmodel/tmodel.go:
```diff
@@ -284,3 +262,99 @@ func ValidateTaskUserName(userName string) error {
+1."queue" is a pointer to the channel which acts as the task queue
+2."tick" is of type Tick struct.
+*/
+func ProcessTaskQueue(queue *chan *Task, tick *Tick) {
```
which code invokes this function? Why do we have to get DB pool/conn here ?

AngithCJ replied:
It is called from here. https://github.com/AngithCJ/ODIM/pull/34/files#diff-18e3404157e24adfa7e24a951819248e66734e443863894985bd516245964e90R48

We are getting a DB pool in every service like this. I also needed to get DB conn here because I needed the same connection to start the pipeline, do the operation and commit the pipeline

Bharath-KKB commented on svc-task/tqueue/queue_test.go:
```diff
@@ -0,0 +1,106 @@
+	"github.com/satori/uuid"
+)
+
+func flushDB(t *testing.T) {
```
why are we doing this ? The tests are all mocks isnt it ?

AngithCJ replied:
No. The DB operations are not mocked.

AngithCJ replied:
For the unit test, we are using a test DB and we are querying this DB to validate the operation. Here I have changed the unit test and DB was not needed for the test. So this function is removed.

Bharath-KKB reviewed (changes requested):
Address comments

AngithCJ committed:
1. New db connection is created once the existing connection is not available
2. Error handling is done
3. Retry the redis transaction once a batch is failed
4. code review changes

AngithCJ committed:
code review changes

AngithCJ committed:
Fixed issues in retry operation for task update

AngithCJ committed:
unit test updated for task queue

Bharath-KKB commented on svc-task/tmodel/tmodel.go:
```diff
@@ -284,3 +278,116 @@ func ValidateTaskUserName(userName string) error {
+		task := dequeueTask(queue)
+
+		if task != nil {
+			saveID := table + ":" + task.ID
```
using integer(taskid) and not string(task URL) here will speed things up. As integer key hashing will be faster than strings. You can add the table and : before sending to DB.

AngithCJ replied:
task.ID is UUID. converting that to an integer will throw an error.

Bharath-KKB commented on svc-task/tmodel/tmodel.go:
```diff
@@ -284,3 +278,116 @@ func ValidateTaskUserName(userName string) error {
+	if len(tasks) > 0 {
+		for i < maxRetry {
+			if err := conn.UpdateTransaction(tasks); err != nil {
+				if err.ErrNo() == errors.DBUpdateFailed {
```
check what is the error and recreate conn only if the original error is related to conn. In case of irrecoverable errors log the full data and return an error.

AngithCJ replied:
I have checked the library code and documentation. There the errors are not categorized. They are returning errors of error type.

Bharath-KKB commented on svc-task/tmodel/tmodel.go:
```diff
@@ -284,3 +278,116 @@ func ValidateTaskUserName(userName string) error {
+	}
+
+	if len(tasks) > 0 {
+		for i < maxRetry {
```
what if there is a set of tasks that were updated and some that didnt get updated ?

AngithCJ replied:
Sorry If did not understand it correctly. we are committing the pipelined transaction after every task in the map is buffered to the pipe. So if any operation fails then we will discard the pipeline and try that whole batch again

Bharath-KKB reviewed (changes requested):
address comments shared earlier and those added in this PR

AngithCJ committed:
code review changes

AngithCJ committed:
Added comments

AngithCJ committed:
fixed lint errors

AngithCJ committed:
unit test failure fixed

AngithCJ committed:
unit test failure fixed

AngithCJ committed:
code cleanup, fixed unit test failure

AngithCJ committed:
map deallocation after processing is done

AngithCJ committed:
1. Implemented validation of retriable Redis errors and retry if it is retriable
2. Unit test improvement

AngithCJ committed:
Fixed lint errors

AngithCJ committed:
Code review changes

AngithCJ committed:
code review changes
